MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -141,6 +151,40 @@ RemoveMemoryMapEntry (
   }
 }
 
+EFI_MEMORY_TYPE
+GetBucketMemoryType (
+  IN EFI_PHYSICAL_ADDRESS  PhysicalStart,
+  IN EFI_PHYSICAL_ADDRESS  PhysicalEnd
+  )
+{
+  EFI_MEMORY_TYPE  BucketType;
+
+  // Find the bucket type for the incoming memory region.
+  for (BucketType = 0; BucketType < EfiMaxMemoryType; BucketType++) {
+    //
+    // If the number of pages for this memory type is not zero, the input region
+    // better be within the same bucket.
+    //
+    if (mMemoryTypeStatistics[BucketType].Special && (mMemoryTypeStatistics[BucketType].NumberOfPages != 0)) {
+      if ((PhysicalStart >= mMemoryTypeStatistics[BucketType].BaseAddress) &&
+          (PhysicalEnd <= mMemoryTypeStatistics[BucketType].MaximumAddress)) {
+        break;

```

Comment:
Should we add an assert here if the memory region partially intersects a bucket but isn't fully contained?

Reply:
I added the assert in the get memory map call instead, which I think is where it really matters

Reply:
It would be good to catch here too, though, right? Because this is where things could go wrong, whereas the get memory map call is when things are already messed up.

Reply:
I think you should just put the condition in a "memory bucket validation" function and call it from both places.